### PR TITLE
Document free memory pointer overflow check in Sol->Yul.

### LINKED ISSUE #11709: [Codegen] Legacy codegen does not check for freememptr overflow (beyond UINT64_MAX), Sol->Yul does
```
pragma abicoder v2;
contract C {
function f() public {
    uint[][] memory arr = new uint[][](1);
    // This allocation overflows the freememptr
    // which makes Sol->Yul panic with code 0x41.
    arr[0] = new uint[](18446744073709551612);
}
}
// ====
// compileViaYul: also
// ----
// f() -> FAILURE
```

Legacy failure is due to out-of-gas, Sol->Yul is a revert with a modest amount of gas consumed ~21k.

### PR DESCRIPTION
**Pull request by bshastry**
Fixes #11709

### CODE REVIEW

File: docs/ir/ir-breaking-changes.rst [hunk 1]
```diff
@@ -193,6 +193,32 @@ This causes differences in some contracts, for example:
   - Old code generator: ``aMod = 0`` and ``mMod = 2``
   - New code generator: ``aMod = 4`` and ``mMod = 0``
 
+- The old code generator does not revert when an impending memory allocation exceeds the free memory pointer
+upper bound of ``type(uint64).max`` (``0xffffffffffffffff``).
+  The new code generator reverts in this scenario.
```
Comment (chriseth) [inline docs/ir/ir-breaking-changes.rst]:
```suggestion
- The new code generator imposes a hard limit of ``type(uint64).max`` (``0xffffffffffffffff``) for the free memory pointer. Allocations that would increase its value beyond this limit revert. The old code generator does not have this limit.
```

File: docs/ir/ir-breaking-changes.rst [hunk 2]
```diff
@@ -193,6 +193,32 @@ This causes differences in some contracts, for example:
...
+              // assumes freeMemPtr points to 0x80 initially
+              uint solYulMaxAllocationBeforeMemPtrOverflow = (type(uint64).max - 0x80 - 31) / 32;
+              // freeMemPtr overflows UINT64_MAX
+              arr = new uint[](solYulMaxAllocationBeforeMemPtrOverflow);
+          }
+      }
+
+  The function `f()` behaves as follows:
+
+  - Old code generator: runs out of gas due to the large memory allocation due to array expansion
```
Comment (chriseth) [inline docs/ir/ir-breaking-changes.rst]:
```suggestion
  - Old code generator: runs out of gas while zeroing the array contents after the large memory allocation
```

File: docs/ir/ir-breaking-changes.rst [hunk 3]
```diff
@@ -193,6 +193,30 @@ This causes differences in some contracts, for example:
...
+      // SPDX-License-Identifier: GPL-3.0
+      pragma solidity >0.8.0;
+      contract C {
+          function f() public {
+              uint[] memory arr;
+              // allocation size: 576460752303423481
+              // assumes freeMemPtr points to 0x80 initially
+              uint solYulMaxAllocationBeforeMemPtrOverflow = (type(uint64).max - 0x80 - 31) / 32;
+              // freeMemPtr overflows UINT64_MAX
+              arr = new uint[](solYulMaxAllocationBeforeMemPtrOverflow);
```
Comment (hrkrshnn) [inline docs/ir/ir-breaking-changes.rst:211]: The optimized code for this function is:

```
                    if eq(0x26121ff0, shr(224, calldataload(_1)))
                    {
                        if callvalue() { revert(_1, _1) }
                        if slt(add(calldatasize(), not(3)), _1) { revert(_1, _1) }
                        mstore(_1, shl(224, 0x4e487b71))
                        mstore(4, 0x41)
                        revert(_1, 0x24)
                    }
```

**Review by hrkrshnn — APPROVED** [review]

File: docs/ir/ir-breaking-changes.rst (hunk 3 above)
Comment (bshastry) [reply docs/ir/ir-breaking-changes.rst:211]: That's nice. Constant propagation I guess. Optimized code would look bigger if length of the dynamic array is a function parameter.

**Pull request merged**